runtime/recipe.js

```diff
@@ -75,7 +75,34 @@ class Recipe {
 
   instantiate(arc) {
     this.beforeInstantiation.forEach(f => f(arc));
-    this.components.forEach(component => component.instantiate(arc));
+    this.components.forEach(component => {
+      component.instantiate(arc)
+
+      if (!arc.availableSlotIds) {
```

Comment:
I cannot create this in the constructor because it needs both recipe and arc.
But I refactored the code - it is a bit more verbose now, but (I think) more readable.
It reduces the number of times I scan components (especially I put it inside the forEach loop by mistake, so)